### PR TITLE
👌 IMPROVE: Configuration migrations

### PR DESCRIPTION
**Pull request by chrisjsewell**
Allow for both upgrades and downgrades, and expose this in the CLI.

This will allow a route for user to downgrade their configuration,
if they try a new version of aiida, but then want to return to an old version.

### CODE REVIEW

**Comment by chrisjsewell** [comment]
I haven't added tests for the CLI commands yet @sphuber; will do if you agree they make sense in general

Note, this is a prelude to adding a migration, to "compartmentalize" config required by a backend (`AIIDADB_NAME`, etc)  to a "storage" key

**Comment by sphuber** [comment]
Cheers @chrisjsewell , guess my first question would be what the advantage is of having this a manual operation? Why shouldn't we just do this automatically? Is the reason because certain downgrades may be lossy? But even in that case, if they want to downgrade, they are forced to migrate anyway, and there will be an automatic backup. So maybe an automatic migration is still the easiest for users?

**Comment by chrisjsewell** [comment]
@sphuber how can you automatically downgrade using an old version of aiida, that does not have that migration available 😬
The only way to downgrade, is to use the version of aiida, that has the upgrade. Unless I'm missing something

**Comment by sphuber** [comment]
> @sphuber how can you automatically downgrade using an old version of aiida, that does not have that migration available grimacing The only way to downgrade, is to use the version of aiida, that has the upgrade. Unless I'm missing something

I thought the goal of the downgrade was the following scenario:

- User has `aiida-core==1.0` with config v1 installed
- User upgrades to `aiida-core==2.0` which has config v2. Currently, this is automatically upgraded.
- User downgrades again to `aiida-core==1.0` and now AiiDA will say the config is not compatible

I thought that your description described this scenario, and this would now allow the user to call `verdi config downgrade` to make their config file compatible again. But my question is whether that should be a manual action or whether we should simply automatically downgrade to the compatible version, just as we upgrade it manually. Why would a user ever want to manually down or upgrade to a version that then won't be compatible with the installation anyway?

**Comment by chrisjsewell** [comment]
> User downgrades again to aiida-core==1.0 and now AiiDA will say the config is not compatible

But how can aiida-core v1 downgrade from a future version?

**Comment by chrisjsewell** [comment]
> Edit: nevermind, I just got the point...

Yep 👍

**Comment by sphuber** [comment]
One major problem that I see with all this is that in order to really try a new version of AiiDA, they will have to migrate their database and those are not downgradable, so there is no point in downgrading the config because you won't be able to use the profile anyway

**Comment by chrisjsewell** [comment]
What if you upgrade and create a new profile, then want to downgrade and use an older (unmigrated) profile?

**Comment by sphuber** [comment]
> What if you upgrade and create a new profile, then want to downgrade and use an older (unmigrated) profile?

Fair. I guess that there are some niche use cases, so I am fine with moving forward with this. Just think we should be very clear up front that the fact that this is here, it isn't possible to always arbitrarily migrate back and forth. Will now review the rest of the PR.

File: aiida/cmdline/commands/cmd_config.py [hunk 1]
```diff
@@ -190,3 +192,30 @@ def verdi_config_caching(disabled):
...
+@click.argument('path', required=False, default=None, type=click.Path(exists=True, path_type=Path))
+@click.option('--version', type=int, default=None, help='Upgrade to specific version (default current).')
+def verdi_config_upgrade(version, path):
+    """Print a configuration, upgraded to a specific version."""
+    from aiida.manage.configuration import CURRENT_CONFIG_VERSION, get_config_path, upgrade_config
+    path = path or Path(get_config_path())
+    version = version if version is not None else CURRENT_CONFIG_VERSION
+    echo.echo_report(f'Upgrading configuration to v{version}: {path}')
+    config = json.loads(path.read_text(encoding='utf8'))
+    upgrade_config(config, version)
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_config.py:207]: Will this actually migrate the provided config and save the changes to disk?

File: aiida/manage/configuration/migrations/migrations.py [hunk 2]
```diff
@@ -8,136 +8,282 @@
...
 CURRENT_CONFIG_VERSION = 5
 OLDEST_COMPATIBLE_CONFIG_VERSION = 5
 
 
-class ConfigMigration:
-    """Defines a config migration."""
+class SingleMigration(Protocol):
+    """A single migration of the configuration."""
+    initial: int
+    """The initial configuration version."""
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:34]:
```suggestion
    initial: int  #The initial configuration version.
```

File: aiida/manage/configuration/migrations/migrations.py [hunk 3]
```diff
@@ -8,136 +8,282 @@
...
+    final: int
+    """The final configuration version."""
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:36]:
```suggestion
    final: int  #The final configuration version.
```

File: aiida/manage/configuration/migrations/migrations.py [hunk 4]
```diff
@@ -8,136 +8,282 @@
...
+
 CURRENT_CONFIG_VERSION = 5
 OLDEST_COMPATIBLE_CONFIG_VERSION = 5
 
 
-class ConfigMigration:
-    """Defines a config migration."""
+class SingleMigration(Protocol):
+    """A single migration of the configuration."""
+    initial: int
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:33]: I take it these `initial` and `final` are like the `down_revision` and `revision` in the SQLA migrations? Would it make sense to use the same terminology here?

File: aiida/manage/configuration/migrations/migrations.py [hunk 5]
```diff
@@ -8,136 +8,282 @@
...
 
-def _1_add_profile_uuid(config):
+    The profile uuid will be used as a general purpose identifier for the profile, in
+    for example the RabbitMQ message queues and exchanges.
+    """
+    initial = 0
+    final = 1
+
+    def upgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:57]: Isn't this migration missing `config.setdefault('CONFIG_VERSION', {})['CURRENT'] = self.version`?

File: aiida/manage/configuration/migrations/migrations.py [hunk 6]
```diff
@@ -8,136 +8,282 @@
...
+
+    def downgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:60]: Should the downgrade not _remove_ these keys, since they were added in the upgrade?

File: aiida/manage/configuration/migrations/migrations.py [hunk 7]
```diff
@@ -8,136 +8,282 @@
...
-        profile['PROFILE_UUID'] = uuid4().hex
+    initial = 1
+    final = 2
 
-    return config
+    def upgrade(self, config: ConfigType) -> None:
+        from uuid import uuid4  # we require this import here, to patch it in the tests
+        for profile in config.get('profiles', {}).values():
+            profile.setdefault('PROFILE_UUID', uuid4().hex)
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:78]: Why is this line necessary? That is from the previous migration no?

File: aiida/manage/configuration/migrations/migrations.py [hunk 8]
```diff
@@ -8,136 +8,282 @@
...
 
+    def downgrade(self, config: ConfigType) -> None:
+        # leave the uuid present, so we could migrate back up
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:82]: I see the point in not actually reverting the upgrade, because you would lose the UUID and if you upgrade again you get a different one. But isn't this a problem with the concept? Maybe in this case the older version will accept an additional key that its schema doesn't know about, but this may not be the case in general, would it?

File: aiida/manage/configuration/migrations/migrations.py [hunk 9]
```diff
@@ -8,136 +8,282 @@
...
+        ]
+
+        for profile in config.get('profiles', {}).values():
+            for key, default in defaults:
+                if key not in profile:
+                    profile[key] = default
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 3
 
+    def downgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 3
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:139]: Again, this is not really a downgrade since it does not undo the upgrade. Could this cause issues if actually loading the downgrade config with an older version of `aiida-core`?

File: aiida/manage/configuration/migrations/migrations.py [hunk 10]
```diff
@@ -8,136 +8,282 @@
...
+        current = get_current_version(config)
+        try:
+            migrator = next(m for m in _MIGRATION_LOOKUPS if m.initial == current)
+        except StopIteration:
+            raise exceptions.ConfigurationError(f'No migration found to upgrade version {current}')
+        if migrator in used:
+            raise exceptions.ConfigurationError(f'Circular migration detected, upgrading to {target}')
+        used.append(migrator)
+        migrator().upgrade(config)
+        config.setdefault('CONFIG_VERSION', {})['CURRENT'] = current = migrator.final
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:230]: Hmm, never seen this syntax before. Wonder how common this is and how many developers know this? For clarity would, the following be easier to read?
```suggestion
        current = migrator.final
        config.setdefault('CONFIG_VERSION', {})['CURRENT'] = current
```

**Review by sphuber — CHANGES_REQUESTED** [review]

File: aiida/cmdline/commands/cmd_config.py [hunk 11]
```diff
@@ -190,3 +192,30 @@ def verdi_config_caching(disabled):
                     echo.echo(identifier)
             elif disabled:
                 echo.echo(identifier)
+
+
+@verdi_config.command('upgrade')
+@click.argument('path', required=False, default=None, type=click.Path(exists=True, path_type=Path))
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply aiida/cmdline/commands/cmd_config.py:198]: Actually I think I will remove this command since, on calling the command, verdi is already auto-upgrading, rendering it kinda useless.
The main thing is the downgrad command

File: aiida/cmdline/commands/cmd_config.py [hunk 12]
```diff
@@ -190,3 +192,30 @@ def verdi_config_caching(disabled):
...
+@click.argument('version', type=int)
+@click.argument('path', required=False, default=None, type=click.Path(exists=True, path_type=Path))
+def verdi_config_downgrade(version, path):
+    """Print a configuration, downgraded to a specific version."""
+    from aiida.manage.configuration import downgrade_config, get_config_path
+    path = path or Path(get_config_path())
+    echo.echo_report(f'Downgrading configuration to v{version}: {path}')
+    config = json.loads(path.read_text(encoding='utf8'))
+    downgrade_config(config, version)
+    print(json.dumps(config, indent=2))
```
Comment (chrisjsewell) [inline aiida/cmdline/commands/cmd_config.py:221]: this does not currently actually save the changes to disk.
Could do, obviously just extra code

File: aiida/manage/configuration/migrations/migrations.py (hunk 4 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:33]: They are indeed, will do

File: aiida/manage/configuration/migrations/migrations.py (hunk 2 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:34]: No, you shouldn't do it like this; triple-back tick comments, under a variable, are recognised correctly as its description, and shown for example in sphinx documentation and LSPs:

image

File: aiida/manage/configuration/migrations/migrations.py (hunk 3 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:36]: see above

File: aiida/manage/configuration/migrations/migrations.py (hunk 5 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:57]: No, this is handled centrally in `upgrade_config`/`downgrade_config`.
Actually I'll add in similar logic for `OLDEST_COMPATIBLE`

File: aiida/manage/configuration/migrations/migrations.py (hunk 8 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:82]: It is currently, and I think it should be enforced, that it always the case that schema allow for additional keys (this is the default in json schema), to allow for these "loseless" upgrades/downgrades.

In #5320, I have also removed the fact that `Profile` strips unknown keys, since I don't see any reason for it

File: aiida/manage/configuration/migrations/migrations.py (hunk 10 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:230]: Sure, if you want me to cater to you lesser programmers 😜

**Comment by chrisjsewell** [comment]
> it isn't possible to always arbitrarily migrate back and forth

why though? I don't see any reason why not

**Comment by chrisjsewell** [comment]
I think I've answered all your comments @sphuber, there is just the outstanding issue of the `verdi config downgrade` functionality+testing, will wait on your comments first though

File: aiida/manage/configuration/migrations/migrations.py (hunk 10 above)
Comment (sphuber) [reply aiida/manage/configuration/migrations/migrations.py]: Just a question of readability in my opinion

Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py]: only pulling your leg 😉, but yeh its changed

File: aiida/manage/configuration/migrations/migrations.py (hunk 5 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py]: (done)

File: aiida/cmdline/commands/cmd_config.py (hunk 12 above)
Comment (sphuber) [reply aiida/cmdline/commands/cmd_config.py:207]: That is the behavior that I would expect though. Was your idea that people call this function and then manually copy past it into their config file? If that is the main use case, what is the advantage of making it indirect and forcing a user to perform a non-trivial manual operation?

**Comment by sphuber** [comment]
> why though? I don't see any reason why not

Maybe in theory it could be possible, but it is not clear that we always can. The config may be more lenient with allowing additional keys, and so having downgrades not being a full undo of an upgrade, but to me it is not clear that with the PostgreSQL database we have the same luxury. Besides, it would mean an even greater development load by forcing us to implement and test the downgrade of each migration, which would also be non-trivial. In any case, the majority of migrations at this moment don't have downgrades, so those would have to be written first. But I really don't think we should go down this road.

File: aiida/manage/configuration/migrations/migrations.py (hunk 2 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py]: I see your PEP8 and raise you https://www.python.org/dev/peps/pep-0224/

Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py]: > "String literals occurring immediately after a simple assignment at the top level of a module, class, or `__init__`` method are called "attribute docstrings"."

Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py]: so yeh, we should be using it

**Comment by chrisjsewell** [comment]
> the majority of migrations at this moment don't have downgrades

well, I'd not all of the sqlalchemy migrations have at least some form of downgrade, which we rely on for the migration testing (you have to migrate down, before you can migrate up)

but no I'm not suggesting we put much more effort inot supporting this

File: aiida/cmdline/commands/cmd_config.py (hunk 12 above)
Comment (chrisjsewell) [reply aiida/cmdline/commands/cmd_config.py:207]: no I was just being lazy on the first parse lol, but will look into now implementing the save to disk

File: aiida/manage/configuration/migrations/migrations.py (hunk 2 above)
Comment (sphuber) [reply aiida/manage/configuration/migrations/migrations.py]: That phrase is the only reference to it and they don't even give an example. There was also a good reason to reject the concept, because it is just not clear to which attribute the docstring belongs to. This is also the reason I stumbled over it:
```python
initial: int
"""Some docstring"""
final: int
```
It is just not logical to think that the docstring belongs to `initial` and not `final`.

Guess in the end it is just personal preference and here I would again consider readability. The fact that certain software tools have added support for it, should not always be sufficient reason to adopt it if the majority of times that the code is read, is without interpretation. But I will let others give their opinion and not fight this if others think it is fine.

Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py]: Its implemented in sphinx and most every Python LSP, it's not just personnel preference. 

> If the majority of times that the code is read, is without interpretation.

I disagree this is the case, and I would certainly recommend you to use  a modern LSP in your development

File: aiida/manage/configuration/migrations/migrations.py (hunk 9 above)
Comment (sphuber) [reply aiida/manage/configuration/migrations/migrations.py]: Fair enough, happy to go ahead with that behavior if it will be guaranteed to work and we assume that behavior from now on

File: aiida/manage/configuration/migrations/migrations.py (hunk 8 above)
Comment (sphuber) [reply aiida/manage/configuration/migrations/migrations.py]: As commented below, fine to go with this behavior from now on and simply condone unknown/superfluous keys

File: aiida/manage/configuration/migrations/migrations.py (hunk 2 above)
Comment (sphuber) [reply aiida/manage/configuration/migrations/migrations.py]: I agree, but in those cases there is no room for confusion since there are never two consecutive declarations without any whitespace. I think it is that then that distinguishes the cases really, at least put a line of whitespace between the docstring and the next attribute declaration. This should make it easier to read and then I am fine with it.

Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py]: Yep fair 👍

File: aiida/cmdline/commands/cmd_config.py (hunk 12 above)
Comment (chrisjsewell) [reply aiida/cmdline/commands/cmd_config.py]: done

**Comment by chrisjsewell** [comment]
Ok this should be good to go @sphuber

File: aiida/manage/configuration/migrations/migrations.py [hunk 13]
```diff
@@ -8,136 +8,299 @@
...
 
+    def upgrade(self, config: ConfigType) -> None:
+        """Migrate the configuration in-place."""
 
-def _1_add_profile_uuid(config):
+    def downgrade(self, config: ConfigType) -> None:
+        """Downgrade the configuration in-place."""
+
+
+class Initial(SingleMigration):
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:53]: Docstring is incorrect

File: aiida/cmdline/commands/cmd_config.py [hunk 14]
```diff
@@ -190,3 +194,31 @@ def verdi_config_caching(disabled):
                     echo.echo(identifier)
             elif disabled:
                 echo.echo(identifier)
+
+
+@verdi_config.command('downgrade')
+@click.argument('version', type=int)
+@click.option(
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_config.py:201]: Correct me if I am wrong, but I don't think you answered my question yet why we need the option to specify an arbitrary input and output file. Since this is user facing it should simply always operate on the currently configured config file. I would suggest to remove this additional complexity since I don't think it is needed for now and you don't have tests for it.

File: aiida/cmdline/commands/cmd_config.py [hunk 15]
```diff
@@ -190,3 +194,31 @@ def verdi_config_caching(disabled):
                     echo.echo(identifier)
             elif disabled:
                 echo.echo(identifier)
+
+
+@verdi_config.command('downgrade')
+@click.argument('version', type=int)
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_config.py:200]: Can we make this a `type=click.Choice` and get the available numbers dynamically from the available migrations? That would make it a lot better from a UI perspective.

File: aiida/manage/configuration/migrations/migrations.py [hunk 16]
```diff
@@ -8,136 +8,299 @@
...
+
+    return config
+
+
+def config_needs_migrating(config, filepath: Optional[str] = None):
+    """Checks if the config needs to be migrated.
+
+    If the oldest compatible version of the configuration is higher than the current configuration version defined
+    in the code, the config cannot be used and so the function will raise.
+
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:291]: Please add docstring for `filepath`

File: aiida/manage/configuration/migrations/migrations.py [hunk 17]
```diff
@@ -8,136 +8,299 @@
...
+    up_compatible = 0
+
+    def upgrade(self, config: ConfigType) -> None:
+        from uuid import uuid4  # we require this import here, to patch it in the tests
+        for profile in config.get('profiles', {}).values():
+            profile.setdefault('PROFILE_UUID', uuid4().hex)
+
+    def downgrade(self, config: ConfigType) -> None:
+        # leave the uuid present, so we could migrate back up
+        pass
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:93]:
```suggestion
```

File: aiida/manage/configuration/migrations/migrations.py [hunk 18]
```diff
@@ -8,136 +8,299 @@
...
+    :return: the migrated configuration dictionary
+    """
+    current = get_current_version(config)
+    used = []
+    while current < target:
+        current = get_current_version(config)
+        try:
+            migrator = next(m for m in _MIGRATION_LOOKUPS if m.down_revision == current)
+        except StopIteration:
+            raise exceptions.ConfigurationError(f'No migration found to upgrade version {current}')
```
Comment (sphuber) [inline aiida/manage/configuration/migrations/migrations.py:237]: Think it is important to add a test for the two error conditions (migration not found and circular dependency) in both upgrade and downgrade.

**Review by sphuber — CHANGES_REQUESTED** [review]

File: aiida/manage/configuration/migrations/migrations.py (hunk 17 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:93]: you can't leave just a comment (and the docstring is being inherited)

File: aiida/manage/configuration/migrations/migrations.py (hunk 16 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:285]: done

File: aiida/manage/configuration/migrations/migrations.py (hunk 13 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:53]: done

File: aiida/manage/configuration/migrations/migrations.py (hunk 18 above)
Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:231]: done

Comment (chrisjsewell) [reply aiida/manage/configuration/migrations/migrations.py:231]: (see tests/manage/configuration/migrations/test_migrations.py)

File: aiida/cmdline/commands/cmd_config.py (hunk 15 above)
Comment (chrisjsewell) [reply aiida/cmdline/commands/cmd_config.py]: done

File: aiida/cmdline/commands/cmd_config.py (hunk 14 above)
Comment (chrisjsewell) [reply aiida/cmdline/commands/cmd_config.py]: removed

**Review by sphuber — APPROVED** [review]
Thanks @chrisjsewell . I am approving so we can merge this, but I did still wanted to note that because the writing of the config in the `downgrade` command now goes outside of the `Config` class, there will be no automatic backup created. Maybe this is something we want to add in the future to prevent losing a configuration if the downgrade were to ever be incorrect.
